File: PiPPlayerWithObjc/PiPPlayerWithObjc/Controller/PiPAVPlayerViewController.m
//
//  PiPAVPlayerViewController.m
//  PiPPlayerWithObjc
//
//  Created by 김혜리 on 2022/04/25.
//

#import "PiPAVPlayerViewController.h"
#import <AVKit/AVKit.h>

@interface PiPAVPlayerViewController () <
AVPictureInPictureControllerDelegate
>

@property (nonatomic, retain) AVPictureInPictureController *pipController;

@end

@implementation PiPAVPlayerViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    // Do any additional setup after loading the view.

    [self settingUpPiPController];
}

- (void)settingUpPiPController {
    NSURL *url = [NSURL URLWithString: self.urlString];
    AVPlayer *player = [[AVPlayer alloc] initWithURL: url];
    AVPlayerLayer *playerLayer = [[AVPlayerLayer alloc] init];
    [playerLayer setFrame: self.view.bounds];
    [playerLayer setPlayer: player];
    [self.view.layer addSublayer: playerLayer];
    [player play];

    if([AVPictureInPictureController isPictureInPictureSupported]){
        self.pipController = [[AVPictureInPictureController alloc] initWithPlayerLayer: playerLayer];
        self.pipController.delegate = self;
        self.pipController.canStartPictureInPictureAutomaticallyFromInline = YES;
    } else {
        NSLog(@"not supported");
    }
}

// MARK: - AVPictureInPictureControllerDelegate for AVPlayer
- (void)pictureInPictureControllerWillStartPictureInPicture:(AVPictureInPictureController *)pictureInPictureController {
    NSLog(@"pictureInPictureControllerWillStartPictureInPicture");
}

- (void)pictureInPictureControllerDidStartPictureInPicture:(AVPictureInPictureController *)pictureInPictureController {
    NSLog(@"pictureInPictureControllerDidStartPictureInPicture");
}

- (void)pictureInPictureController:(AVPictureInPictureController *)pictureInPictureController failedToStartPictureInPictureWithError:(NSError *)error {
    NSLog(@"failedToStartPictureInPictureWithError");
}

- (void)pictureInPictureController:(AVPictureInPictureController *)pictureInPictureController restoreUserInterfaceForPictureInPictureStopWithCompletionHandler:(void (^)(BOOL))completionHandler {
    NSLog(@"restoreUserInterfaceForPictureInPictureStopWithCompletionHandler");
}

- (void)pictureInPictureControllerWillStopPictureInPicture:(AVPictureInPictureController *)pictureInPictureController {
    NSLog(@"pictureInPictureControllerWillStopPictureInPicture");
}

- (void)pictureInPictureControllerDidStopPictureInPicture:(AVPictureInPictureController *)pictureInPictureController {
    NSLog(@"pictureInPictureControllerDidStopPictureInPicture");
}

@end

File: PiPPlayerWithSwift/PiPPlayerWithSwift/Controller/PiPAVPlayerViewController.swift
//
//  PiPAVPlayerViewController.swift
//  PiPPlayerWithSwift
//
//  Created by 김혜리 on 2022/04/24.
//

import UIKit
import AVKit

final class PiPAVPlayerViewController: UIViewController {
    
    private var pipController: AVPictureInPictureController?
    var urlString: String?
    
    override func viewDidLoad() {
        super.viewDidLoad()
        
        settingUpPiPController()
    }
    
    private func settingUpPiPController() {
        guard let urlString = urlString,
              let url = URL(string: urlString) else {
            return
        }
        
        let player = AVPlayer(url: url)
        let playerLayer = AVPlayerLayer(player: player)
        playerLayer.frame = self.view.bounds
        self.view.layer.addSublayer(playerLayer)
        player.play()
        
        if AVPictureInPictureController.isPictureInPictureSupported() {
            pipController = AVPictureInPictureController(playerLayer: playerLayer)
            pipController?.delegate = self
            pipController?.canStartPictureInPictureAutomaticallyFromInline = true
        } else {
            print("not supported")
        }
    }
}

// MARK: - AVPictureInPictureControllerDelegate
extension PiPAVPlayerViewController: AVPictureInPictureControllerDelegate {
    func pictureInPictureControllerWillStartPictureInPicture(_ pictureInPictureController: AVPictureInPictureController) {
        print("pictureInPictureControllerWillStartPictureInPicture")
    }
    
    func pictureInPictureControllerDidStartPictureInPicture(_ pictureInPictureController: AVPictureInPictureController) {
        print("pictureInPictureControllerDidStartPictureInPicture")
    }
    
    func pictureInPictureController(_ pictureInPictureController: AVPictureInPictureController, failedToStartPictureInPictureWithError error: Error) {
        print("failedToStartPictureInPictureWithError")
    }
    
    func pictureInPictureControllerWillStopPictureInPicture(_ pictureInPictureController: AVPictureInPictureController) {
        print("pictureInPictureControllerWillStopPictureInPicture")
    }
    
    func pictureInPictureControllerDidStopPictureInPicture(_ pictureInPictureController: AVPictureInPictureController) {
        print("pictureInPictureControllerDidStopPictureInPicture")
    }
    
    func pictureInPictureController(_ pictureInPictureController: AVPictureInPictureController, restoreUserInterfaceForPictureInPictureStopWithCompletionHandler completionHandler: @escaping (Bool) -> Void) {
        print("restoreUserInterfaceForPictureInPictureStopWithCompletionHandler")
    }
}
